### PR TITLE
Implement the notifier of WeCom Robot

### PR DESCRIPTION
**Pull request by magiceses**
## Description

The purpose of this pr is to enable alertmanager to send notifications to [WeCom Chatbot](https://developer.work.weixin.qq.com/document/path/91770#%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8%E7%BE%A4%E6%9C%BA%E5%99%A8%E4%BA%BA). For this purpose, i add a new Notifier and some new configs.It is a example of the wecom robot receiver config.
```yaml
- name: wecombot
  wecomrobot_configs:
  - send_resolved: false
    message: '{{ template "wechat.custom.message" . }}'
    webhook_url: "https://qyapi.weixin.qq.com/cgi-bin/webhook/send?key=xxxxxxxx"
```
image

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Implement the notifier of WeCom Robot
```

### CODE REVIEW

File: cmd/alertmanager/main.go [hunk 1]
```diff
@@ -181,6 +182,9 @@ func buildReceiverIntegrations(nc *config.Receiver, tmpl *template.Template, log
 	for i, c := range nc.WebexConfigs {
 		add("webex", i, c, func(l log.Logger) (notify.Notifier, error) { return webex.New(c, tmpl, l) })
 	}
+	for i, c := range nc.WeComRobotConfigs {
```
Comment (hoperays) [inline cmd/alertmanager/main.go]: Leave a blank line to facilitate merging with other patches or upgrade versions.

File: config/config.go [hunk 2]
```diff
@@ -254,6 +254,9 @@ func resolveFilepaths(baseDir string, cfg *Config) {
 		for _, cfg := range receiver.WebexConfigs {
 			cfg.HTTPConfig.SetDirectory(baseDir)
 		}
+		for _, cfg := range receiver.WeComRobotConfigs {
```
Comment (hoperays) [inline config/config.go]: ditto.

File: config/config.go [hunk 3]
```diff
@@ -528,6 +531,14 @@ func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
 				webex.APIURL = c.Global.WebexAPIURL
 			}
 		}
+		for _, wecomrobot := range rcv.WeComRobotConfigs {
```
Comment (hoperays) [inline config/config.go]: ditto.

File: config/config.go [hunk 4]
```diff
@@ -884,18 +895,19 @@ type Receiver struct {
...
+	SlackConfigs      []*SlackConfig      `yaml:"slack_configs,omitempty" json:"slack_configs,omitempty"`
+	WebhookConfigs    []*WebhookConfig    `yaml:"webhook_configs,omitempty" json:"webhook_configs,omitempty"`
+	OpsGenieConfigs   []*OpsGenieConfig   `yaml:"opsgenie_configs,omitempty" json:"opsgenie_configs,omitempty"`
+	WechatConfigs     []*WechatConfig     `yaml:"wechat_configs,omitempty" json:"wechat_configs,omitempty"`
+	PushoverConfigs   []*PushoverConfig   `yaml:"pushover_configs,omitempty" json:"pushover_configs,omitempty"`
+	VictorOpsConfigs  []*VictorOpsConfig  `yaml:"victorops_configs,omitempty" json:"victorops_configs,omitempty"`
+	SNSConfigs        []*SNSConfig        `yaml:"sns_configs,omitempty" json:"sns_configs,omitempty"`
+	TelegramConfigs   []*TelegramConfig   `yaml:"telegram_configs,omitempty" json:"telegram_configs,omitempty"`
+	WebexConfigs      []*WebexConfig      `yaml:"webex_configs,omitempty" json:"webex_configs,omitempty"`
+	WeComRobotConfigs []*WeComRobotConfig `yaml:"wecomrobot_configs,omitempty" json:"wecomrobot_configs,omitempty"`
```
Comment (hoperays) [inline config/config.go]: ditto.

File: config/notifiers.go [hunk 5]
```diff
@@ -164,6 +164,14 @@ var (
 		Message:              `{{ template "telegram.default.message" . }}`,
 		ParseMode:            "HTML",
 	}
+
+	// DefaultWeComRobotConfig defines default values for WeComRobot configurations.
+	DefaultWeComRobotConfig = WeComRobotConfig{
+		NotifierConfig: NotifierConfig{
+			VSendResolved: true,
```
Comment (hoperays) [inline config/notifiers.go]: True or False?

File: notify/wecomrobot/wecomrobot.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
+package wecomrobot
```
Comment (hoperays) [inline notify/wecomrobot/wecomrobot.go:14]: Lack of a header with license info.

File: notify/wecomrobot/wecomrobot.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+type WeComRobotMessage struct {
+	Msgtype string `json:"msgtype"`
+	Text    Mark   `json:"text"`
+}
+
+// Notify implements the Notifier interface.
+func (n *Notifier) Notify(ctx context.Context, alerts ...*types.Alert) (bool, error) {
+	fmt.Println("alert:", alerts[0].Labels)
```
Comment (hoperays) [inline notify/wecomrobot/wecomrobot.go]: For debug?

File: notify/wecomrobot/wecomrobot.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	}
+	defer notify.Drain(resp)
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return true, err
+	}
+	level.Debug(n.logger).Log("response", string(body))
+
+	return true, nil
```
Comment (hoperays) [inline notify/wecomrobot/wecomrobot.go]: Is it necessary to parse fields of resp.Body to check success or failure of internal business logic?

File: config/config.go [hunk 9]
```diff
@@ -529,6 +533,15 @@ func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
 			}
 		}
 
+		for _, wecomrobot := range rcv.WeComRobotConfigs {
+			if wecomrobot.HTTPConfig == nil {
+				wecomrobot.HTTPConfig = c.Global.HTTPConfig
+			}
+			if wecomrobot.WebhookURL == nil {
+				return fmt.Errorf("no wecombot URL provided")
+			}
```
Comment (hoperays) [inline config/config.go]: Duplicate with the check in UnmarshalYAML()?

File: config/notifiers.go [hunk 10]
```diff
@@ -204,6 +212,32 @@ func (c *WebexConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
 
+// WeComRobotConfig configures notifications via WeCom.
+type WeComRobotConfig struct {
+	NotifierConfig `yaml:",inline" json:",inline"`
+	HTTPConfig     *commoncfg.HTTPClientConfig `json:"httpConfig,omitempty"`
+
+	// Template to generate the message content of the robot notification.
+	Message string `yaml:"message,omitempty" json:"message,omitempty"`
+	// The webhook URL of robot which the message will send to.
+	WebhookURL *SecretURL `yaml:"webhook_url,omitempty" json:"webhook_url,omitempty"`
```
Comment (hoperays) [inline config/notifiers.go]: Since the webhook_url field is required, the omitempty tag should be removed.

File: config/notifiers.go [hunk 11]
```diff
@@ -204,6 +212,32 @@ func (c *WebexConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *WeComRobotConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultWeComRobotConfig
+	type plain WeComRobotConfig
+	if err := unmarshal((*plain)(c)); err != nil {
+		return err
+	}
+
+	if c.WebhookURL == nil || c.WebhookURL.URL.String() == "" {
```
Comment (hoperays) [inline config/notifiers.go]: Abbreviated to c.WebhookURL.String()?

File: notify/wecomrobot/wecomrobot.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+		tmpl    = notify.TmplText(n.tmpl, data, &tmplErr)
+	)
+	if tmplErr != nil {
+		return false, fmt.Errorf("templating error: %s", tmplErr)
+	}
+
+	message := tmpl(n.conf.Message)
+	if tmplErr != nil {
+		return false, fmt.Errorf("templating error: %s", tmplErr)
+	}
```
Comment (hoperays) [inline notify/wecomrobot/wecomrobot.go:77]: Unnecessary?

Comment (magiceses) [reply notify/wecomrobot/wecomrobot.go:77]: line 73-75 is Unnecessary, deleted

**Comment by hoperays** [comment]
LGTM

File: config/notifiers.go [hunk 13]
```diff
@@ -204,6 +212,32 @@ func (c *WebexConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	return nil
 }
 
+// WeComRobotConfig configures notifications via WeCom Robot.
+type WeComRobotConfig struct {
+	NotifierConfig `yaml:",inline" json:",inline"`
+	HTTPConfig     *commoncfg.HTTPClientConfig `json:"httpConfig,omitempty"`
```
Comment (charlottexl) [inline config/notifiers.go]: 1. Pelase change httpConfig to  http_config.
2. Missing yaml parsing mapping: 
   `yaml:"http_config,omitempty"

**Comment by hoperays** [comment]
LGTM

File: config/notifiers.go [hunk 14]
```diff
@@ -204,6 +213,35 @@ func (c *WebexConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+	NotifierConfig `yaml:",inline" json:",inline"`
+	HTTPConfig     *commoncfg.HTTPClientConfig `yaml:"http_config,omitempty" json:"http_config,omitempty"`
+
+	// Template to generate the message content of the robot notification.
+	Message string `yaml:"message,omitempty" json:"message,omitempty"`
+	// The webhook URL of robot which the message will be sent.
+	WebhookURL *SecretURL `yaml:"webhook_url" json:"webhook_url"`
+	// The maximum message length supported by the robot notification, unit: byte.
+	// And the excess part will be truncated.
+	MaxMessageSize int `yaml:"max_message_size,omitempty" json:"max_message_size,omitempty"`
```
Comment (charlottexl) [inline config/notifiers.go:227]: Does maxMessageSize need to be greater than 0?

**Review by charlottexl — COMMENTED** [review]
LGTM

**Pull request merged**